# 4. Data Manipulation II - Advanced Selecting - Quiz.ipynb

## 4.0. Import `polars` and Load Data

In [1]:
import polars as pl

In [2]:
df = pl.read_parquet("../data/yellow_tripdata_2024-03.parquet")

## 4.1 Question 1: String Columns Count

Select all columns from the DataFrame that have the datatype `pl.String`. How many columns are there in the result?

In [3]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

(3582628, 1)


1. 1
2. 19
3. 3
4. 0

## 4.2 Question 2: Int64 Columns Count

How many columns in the DataFrame have the datatype `pl.Int64`?

In [4]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

(3582628, 3)


1. 19
2. 1
3. 3
4. 3582628

## 4.3 Question 3: Rows with Zero Float64 Values

What fraction of rows have at least one of their `pl.Float64` columns equal to exactly 0? (Hint: Use `pl.any_horizontal()`.)

In [5]:
result = (
    df
    .select(
        #### YOUR CODE HERE
    )
)
print(result)

shape: (1, 1)
┌──────────┐
│ fraction │
│ ---      │
│ f64      │
╞══════════╡
│ 0.0      │
└──────────┘


1. 0
2. 0.000076
3. 0.50
4. 1.0

## 4.4 Question 4: Highest Tip and Fare for New Column

Create a new column called `tip_amount_plus_fare_amount`. Then sort the DataFrame by this new column in descending order. What are the `tip_amount` and `fare_amount` for the highest `tip_amount_plus_fare_amount`?

In [6]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

shape: (3_582_628, 3)
┌────────────┬─────────────┬─────────────────────────────┐
│ tip_amount ┆ fare_amount ┆ tip_amount_plus_fare_amount │
│ ---        ┆ ---         ┆ ---                         │
│ f64        ┆ f64         ┆ f64                         │
╞════════════╪═════════════╪═════════════════════════════╡
│ 999.99     ┆ 17.0        ┆ 1016.99                     │
│ 0.0        ┆ 900.0       ┆ 900.0                       │
│ 0.0        ┆ 850.0       ┆ 850.0                       │
│ 0.0        ┆ 840.0       ┆ 840.0                       │
│ 0.0        ┆ 800.0       ┆ 800.0                       │
│ …          ┆ …           ┆ …                           │
│ 0.0        ┆ -558.8      ┆ -558.8                      │
│ 1.0        ┆ -600.0      ┆ -599.0                      │
│ 0.0        ┆ -619.7      ┆ -619.7                      │
│ 161.2      ┆ -800.0      ┆ -638.8                      │
│ 0.01       ┆ -800.0      ┆ -799.99                     │
└────────────┴─────────────┴──────

1. `tip_amount = 17.0, fare_amount = 999.99`
2. `tip_amount = 0.01, fare_amount = -800.0`
3. `tip_amount = 999.99, fare_amount = 17.0`
4. `tip_amount = 999.99, fare_amount = 999.99`

## 4.5 Question 5: Same Pickup-Dropoff Fraction

What fraction of rides have the same location for pickup and dropoff?

In [7]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

shape: (1, 1)
┌──────────────┐
│ DOLocationID │
│ ---          │
│ f64          │
╞══════════════╡
│ 0.050603     │
└──────────────┘


1. 0.949397
2. 0.050603
3. 1.0
4. 0.0

## 4.6 Question 6: Duplicate Columns with Suffix

Using `.with_columns()`, `pl.all()`, and `.name.suffix()`, add to the DataFrame a copy of all the columns with `_new` added to the end of each column name. How many columns are there in the resultant DataFrame?

In [8]:
result = (
    df
    #### YOUR CODE HERE
)
print(result.shape)

(3582628, 38)


1. 3582628
2. 38
3. 19
4. 22

## 4.7 Question 7: Empty String Check Columns

Add a new column to the DataFrame for every `pl.String` column that checks whether that column has an empty string (i.e., a string equal to `""`). How many columns are in the resultant DataFrame?

In [9]:
result = (
    df
    #### YOUR CODDE HERE
)
print(result)

(3582628, 20)


1. 1
2. 20
3. 22
4. 38

## 4.8 Question 8: Same Location Trips Dataframe

Your goal is to produce a DataFrame of only the rides that took place in one location (i.e., where `DOLocationID` equals `PULocationID`). This means that you no longer need both of the columns `DOLocationID` and `PULocationID` (since they are equal). So filter for trips with the same pickup and dropoff sites and remove either one of the pickup/dropoff location columns. Then rename the remaining one to `LocationID`. What is the shape of the resultant DataFrame?

In [10]:
result = (
    df
    #### YOUR CODE HERE
)
print(result)

(3582628, 18)


1. (181291, 18)
2. (3582628, 18)
3. (181291, 19)
4. (3401337, 18)